lessons/cryptography/cryptography-hashing-checksums-salts.md

```diff
@@ -0,0 +1,82 @@
+
+One way to help prevent the same hash from appearing in a stored list even when multiple users choose the same password is to use a technique called password salting. A salt is a random string of characters that is added to a user's password prior to hashing. Since the salt is always random, this would result in different password hashes even if multiple users choose the same password. Although the salt is a random string of letters, it is only generated randomly the first time when a user initially creates an account. After that, the salt information would be stored along with the username and hashed password, so that when the user signs in in the future, the salt would be added to the password, then hashed, and then compared with the stored value in the list. Adding a salt to every user's password not only helps prevent the same hashes from appearing multiple times, but also takes away the usefulness of the lists of common passwords and their hashes. 
+
+# The Real World, Prolific Breaches
+A good cryptographic hash function should never result in the same hash value from two different messages. To find weaknesses of the cryptographic hash algorithms, people attempt to find two input messages that would produce the same hash value. If this occurs, it is called a collision. The MD5 hash has been shown to have many collisions that are possible. With today's computing power, it is even possible to find collisions within seconds. About 20 years after the SHA-1 hash was introduced, a group of researches at Google and the CWI Institute in Amsterdam showed that they were able to provide two different PDF files that have different content but result in the same SHA-1 hash value. 
+
+# Check YoSelf
+## Q1 - True or False? If you have a hash, you can reverse it to obtain the original message. 
+a. True <br>
+b. False <---<br>
+
+## Q2 - True or False? Cryptographic hashes are often used for password verification because is it not possible to store plaintext passwords. 
+a. True <br>
+b. False <---<br>
+
+## Q3 - True or False? Password salting is used to prevent the same hash from appearing in a list of password hashes if multiple users have the same plaintext password. 
+a. True <---<br>
+b. False <br>
+
+## Q4 - What is the purpose of a checksum?
+a. Check a message for spelling errors. <br>
+b. Verify that a message is not tampered with during message transmission. <---<br>
+c. Verify the identity of the sender of a message. <br>
+d. All of the above. <br>
+
+# Keep Going, Next Steps
```

Comment:
Maybe include https://howsecureismypassword.net/ as an example, but this is already in the password security lesson so its at your discretion.